File: src/main/java/at/bitfire/ical4android/Event.java
/*
 * Copyright (c) 2013 – 2015 Ricki Hirner (bitfire web engineering).
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 */

package at.bitfire.ical4android;

import net.fortuna.ical4j.data.CalendarOutputter;
import net.fortuna.ical4j.data.ParserException;
import net.fortuna.ical4j.model.Calendar;
import net.fortuna.ical4j.model.Component;
import net.fortuna.ical4j.model.ComponentList;
import net.fortuna.ical4j.model.Date;
import net.fortuna.ical4j.model.DateTime;
import net.fortuna.ical4j.model.Property;
import net.fortuna.ical4j.model.PropertyList;
import net.fortuna.ical4j.model.TimeZone;
import net.fortuna.ical4j.model.ValidationException;
import net.fortuna.ical4j.model.component.VAlarm;
import net.fortuna.ical4j.model.component.VEvent;
import net.fortuna.ical4j.model.property.Attendee;
import net.fortuna.ical4j.model.property.Clazz;
import net.fortuna.ical4j.model.property.Description;
import net.fortuna.ical4j.model.property.DtEnd;
import net.fortuna.ical4j.model.property.DtStamp;
import net.fortuna.ical4j.model.property.DtStart;
import net.fortuna.ical4j.model.property.Duration;
import net.fortuna.ical4j.model.property.ExDate;
import net.fortuna.ical4j.model.property.ExRule;
import net.fortuna.ical4j.model.property.LastModified;
import net.fortuna.ical4j.model.property.Location;
import net.fortuna.ical4j.model.property.Organizer;
import net.fortuna.ical4j.model.property.ProdId;
import net.fortuna.ical4j.model.property.RDate;
import net.fortuna.ical4j.model.property.RRule;
import net.fortuna.ical4j.model.property.RecurrenceId;
import net.fortuna.ical4j.model.property.Sequence;
import net.fortuna.ical4j.model.property.Status;
import net.fortuna.ical4j.model.property.Summary;
import net.fortuna.ical4j.model.property.Transp;
import net.fortuna.ical4j.model.property.Uid;
import net.fortuna.ical4j.model.property.Version;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.logging.Level;

import lombok.Cleanup;
import lombok.NonNull;
import lombok.ToString;

@ToString(of={"uid","dtStart","summary"})
public class Event extends iCalendar {
    public static final String CALENDAR_NAME = "X-WR-CALNAME";

    // uid and sequence are inherited from iCalendar
    public RecurrenceId recurrenceId;

    public String summary, location, description;

    public DtStart dtStart;
    public DtEnd dtEnd;

    public Duration duration;
    public RRule rRule;
    public ExRule exRule;
    public final List<RDate> rDates = new LinkedList<>();
    public final List<ExDate> exDates = new LinkedList<>();

    public final List<Event> exceptions = new LinkedList<>();

    public Boolean forPublic;
    public Status status;

    public boolean opaque;

    public Organizer organizer;
    public final List<Attendee> attendees = new LinkedList<>();

    public final List<VAlarm> alarms = new LinkedList<>();

    // unknown properties
    protected static final String[] knownPropertyNames = {
            Uid.UID, Sequence.SEQUENCE,
            RecurrenceId.RECURRENCE_ID,
            Summary.SUMMARY, Location.LOCATION, Description.DESCRIPTION,
            DtStart.DTSTART, DtEnd.DTEND,
            Duration.DURATION, RRule.RRULE, ExRule.EXRULE, RDate.RDATE, ExDate.EXDATE,
            Clazz.CLASS,
            Status.STATUS, Transp.TRANSP,
            Organizer.ORGANIZER, Attendee.ATTENDEE,
            DtStamp.DTSTAMP, LastModified.LAST_MODIFIED     // ignore current DTSTAMP and LAST-MODIFIED
    };
    public final List<Property> unknownProperties = new LinkedList<>();


    /**
     * Parses an InputStream that contains iCalendar VEVENTs.
     *
     * @param stream        input stream containing the VEVENTs
     * @param charset       charset of the input stream or null (will assume UTF-8)
     * @param properties    map of properties, will be filled with CALENDAR_* values, if applicable (may be null)
     * @return              array of filled Event data objects (may have size 0) – doesn't return null
     * @throws InvalidCalendarException on parser exceptions
     */
    @SuppressWarnings("unchecked")
    public static Event[] fromStream(@NonNull InputStream stream, Charset charset, Map<String, String> properties) throws IOException, InvalidCalendarException {
        Constants.log.fine("Parsing iCalendar stream");
        final Calendar ical;
        try {
            if (charset != null) {
                @Cleanup InputStreamReader reader = new InputStreamReader(stream, charset);
                ical = calendarBuilder().build(reader);
            } else
                ical = calendarBuilder().build(stream);
        } catch (ParserException e) {
            throw new InvalidCalendarException("Couldn't parse calendar resource", e);
        }

        if (properties != null) {
            Property calName = ical.getProperty(CALENDAR_NAME);
            if (calName != null)
                properties.put(CALENDAR_NAME, calName.getValue());
        }

        List<VEvent> vEvents = ical.getComponents(Component.VEVENT);

        // make sure every event has an UID
        for (VEvent vEvent : vEvents)
            if (vEvent.getUid() == null) {
                Uid uid = new Uid(UUID.randomUUID().toString());
                Constants.log.warning("Found VEVENT without UID, using a random one: " + uid.getValue());
                vEvent.getProperties().add(uid);
            }

        Constants.log.fine("Assigning exceptions to master events");
        Map<String,VEvent> masterEvents = new HashMap<>(vEvents.size());
        Map<String,Map<String,VEvent>> exceptions = new HashMap<>();

        for (VEvent vEvent : vEvents) {
            final String uid = vEvent.getUid().getValue();
            int sequence = vEvent.getSequence() == null ? 0 : vEvent.getSequence().getSequenceNo();

            if (vEvent.getRecurrenceId() == null) {
                // master event (no RECURRENCE-ID)
                VEvent event = masterEvents.get(uid);
                // If there are multiple entries, compare SEQUENCE and use the one with higher SEQUENCE.
                // If the SEQUENCE is identical, use latest version.
                if (event == null || (event.getSequence() != null && sequence >= event.getSequence().getSequenceNo()))
                    masterEvents.put(uid, vEvent);

            } else {
                // exception (RECURRENCE-ID)
                Map<String,VEvent> ex = exceptions.get(uid);
                // first index level: UID
                if (ex == null) {
                    ex = new HashMap<>();
                    exceptions.put(uid, ex);
                }
                // second index level: RECURRENCE-ID
                String recurrenceID = vEvent.getRecurrenceId().getValue();
                VEvent event = ex.get(recurrenceID);
                if (event == null || (event.getSequence() != null && sequence >= event.getSequence().getSequenceNo()))
                    ex.put(recurrenceID, vEvent);
            }
        }

        List<Event> events = new ArrayList<>(masterEvents.size());
        for (Map.Entry<String, VEvent> masterEvent : masterEvents.entrySet()) {
            String uid = masterEvent.getKey();
            Event event = fromVEvent(masterEvent.getValue());

            Map<String,VEvent> eventExceptions = exceptions.get(uid);
            if (eventExceptions != null)
                for (VEvent ex : eventExceptions.values())
                    event.exceptions.add(fromVEvent(ex));

            events.add(event);
        }

        return events.toArray(new Event[events.size()]);
    }

    /**
     * Same as #{@link #fromStream(InputStream, Charset, Map)}, but without properties map.
     */
    public static Event[] fromStream(@NonNull InputStream stream, Charset charset) throws IOException, InvalidCalendarException {
        return fromStream(stream, charset, null);
    }


    protected static Event fromVEvent(VEvent event) throws InvalidCalendarException {
        final Event e = new Event();

        // default values
        e.sequence = 0;
        e.opaque = true;
        e.forPublic = true;

        // process properties
        for (Property prop : event.getProperties()) {
            if (prop instanceof Uid)
                e.uid = prop.getValue();

            else if (prop instanceof RecurrenceId)
                e.recurrenceId = (RecurrenceId)prop;

            else if (prop instanceof Sequence)
                e.sequence = ((Sequence)prop).getSequenceNo();

            else if (prop instanceof DtStart)
                e.dtStart = (DtStart)prop;
            else if (prop instanceof DtEnd)
                e.dtEnd = (DtEnd)prop;
            else if (prop instanceof Duration)
                e.duration = (Duration)prop;

            else if (prop instanceof RRule)
                e.rRule = (RRule)prop;
            else if (prop instanceof RDate)
                e.rDates.add((RDate)prop);
            else if (prop instanceof ExRule)
                e.exRule = (ExRule)prop;
            else if (prop instanceof ExDate)
                e.exDates.add((ExDate)prop);

            else if (prop instanceof Summary)
                e.summary = prop.getValue();
            else if (prop instanceof Location)
                e.location = prop.getValue();
            else if (prop instanceof Description)
                e.description = prop.getValue();

            else if (prop instanceof Status)
                e.status = (Status)prop;
            else if (prop instanceof Transp)
                e.opaque = prop == Transp.OPAQUE;

            else if (prop instanceof Clazz)
                e.forPublic = prop == Clazz.PUBLIC;

            else if (prop instanceof Organizer)
                e.organizer = (Organizer)prop;
            else if (prop instanceof Attendee)
                e.attendees.add((Attendee)prop);

            else if (prop instanceof ProdId || prop instanceof DtStamp || prop instanceof LastModified)
                /* ignore */;

            else    // retain unknown properties
                e.unknownProperties.add(prop);
        }

        // calculate DtEnd from Duration
        if (e.dtEnd == null && e.duration != null)
            e.dtEnd = event.getEndDate(true);

        e.alarms.addAll(event.getAlarms());

        // validation
        if (e.dtStart == null)
            throw new InvalidCalendarException("Event without start time");

        validateTimeZone(e.dtStart);
        validateTimeZone(e.dtEnd);

        return e;
    }

    @SuppressWarnings("unchecked")
    public void write(OutputStream os) throws IOException {
        net.fortuna.ical4j.model.Calendar ical = new net.fortuna.ical4j.model.Calendar();
        ical.getProperties().add(Version.VERSION_2_0);
        ical.getProperties().add(prodId);

        // "master event" (without exceptions)
        ComponentList components = ical.getComponents();
        VEvent master = toVEvent(new Uid(uid));
        components.add(master);

        // remember used time zones
        Set<TimeZone> usedTimeZones = new HashSet<>();
        if (dtStart != null && dtStart.getTimeZone() != null)
            usedTimeZones.add(dtStart.getTimeZone());
        if (dtEnd != null && dtEnd.getTimeZone() != null)
            usedTimeZones.add(dtEnd.getTimeZone());

        // recurrence exceptions
        for (Event exception : exceptions) {
            // create VEVENT for exception
            VEvent vException = exception.toVEvent(master.getUid());

            components.add(vException);

            // remember used time zones
            if (exception.dtStart != null && exception.dtStart.getTimeZone() != null)
                usedTimeZones.add(exception.dtStart.getTimeZone());
            if (exception.dtEnd != null && exception.dtEnd.getTimeZone() != null)
                usedTimeZones.add(exception.dtEnd.getTimeZone());
        }

        // add VTIMEZONE components
        for (net.fortuna.ical4j.model.TimeZone timeZone : usedTimeZones)
            ical.getComponents().add(timeZone.getVTimeZone());

        CalendarOutputter output = new CalendarOutputter(false);
        try {
            output.output(ical, os);
        } catch (ValidationException e) {
            Constants.log.log(Level.SEVERE, "Couldn't generate valid VEVENT", e);
        }
    }

    protected VEvent toVEvent(Uid uid) {
        VEvent event = new VEvent();
        PropertyList props = event.getProperties();

        if (uid != null)
            props.add(uid);
        if (recurrenceId != null)
            props.add(recurrenceId);
        if (sequence != null && sequence != 0)
            props.add(new Sequence(sequence));

        props.add(dtStart);
        if (dtEnd != null)
            props.add(dtEnd);
        if (duration != null)
            props.add(duration);

        if (rRule != null)
            props.add(rRule);
        for (RDate rdate : rDates)
            props.add(rdate);
        if (exRule != null)
            props.add(exRule);
        for (ExDate exdate : exDates)
            props.add(exdate);

        if (summary != null && !summary.isEmpty())
            props.add(new Summary(summary));
        if (location != null && !location.isEmpty())
            props.add(new Location(location));
        if (description != null && !description.isEmpty())
            props.add(new Description(description));

        if (status != null)
            props.add(status);
        if (!opaque)
            props.add(Transp.TRANSPARENT);

        if (organizer != null)
            props.add(organizer);
        props.addAll(attendees);

        if (forPublic != null)
            event.getProperties().add(forPublic ? Clazz.PUBLIC : Clazz.PRIVATE);

        props.addAll(unknownProperties);

        event.getAlarms().addAll(alarms);
        return event;
    }


    // time helpers

    public boolean isAllDay() {
        return !isDateTime(dtStart);
    }

    public long getDtStartInMillis() {
        return dtStart.getDate().getTime();
    }

    public String getDtStartTzID() {
        return getTzId(dtStart);
    }

    public void setDtStart(long tsStart, String tzID) {
        if (tzID == null) {    // all-day
            dtStart = new DtStart(new Date(tsStart));
        } else {
            DateTime start = new DateTime(tsStart);
            start.setTimeZone(DateUtils.tzRegistry.getTimeZone(tzID));
            dtStart = new DtStart(start);
        }
    }


    public long getDtEndInMillis() {
        return dtEnd.getDate().getTime();
    }

    public String getDtEndTzID() {
        return getTzId(dtEnd);
    }

    public void setDtEnd(long tsEnd, String tzID) {
        if (tzID == null) {    // all-day
            dtEnd = new DtEnd(new Date(tsEnd));
        } else {
            DateTime end = new DateTime(tsEnd);
            end.setTimeZone(DateUtils.tzRegistry.getTimeZone(tzID));
            dtEnd = new DtEnd(end);
        }
    }

}

File: src/main/java/at/bitfire/ical4android/Event.kt
/***************************************************************************************************
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 **************************************************************************************************/

package at.bitfire.ical4android

import at.bitfire.ical4android.ICalendar.Companion.CALENDAR_NAME
import at.bitfire.ical4android.util.DateUtils.isDateTime
import at.bitfire.ical4android.validation.EventValidator
import net.fortuna.ical4j.data.CalendarOutputter
import net.fortuna.ical4j.data.ParserException
import net.fortuna.ical4j.model.*
import net.fortuna.ical4j.model.Calendar
import net.fortuna.ical4j.model.TimeZone
import net.fortuna.ical4j.model.component.VAlarm
import net.fortuna.ical4j.model.component.VEvent
import net.fortuna.ical4j.model.parameter.Email
import net.fortuna.ical4j.model.property.*
import java.io.IOException
import java.io.OutputStream
import java.io.Reader
import java.net.URI
import java.util.*

class Event: ICalendar() {

    /** list of CUAs which have edited the event since last sync */
    var userAgents = LinkedList<String>()

    // uid and sequence are inherited from iCalendar
    var recurrenceId: RecurrenceId? = null

    var summary: String? = null
    var location: String? = null
    var url: URI? = null
    var description: String? = null
    var color: Css3Color? = null

    var dtStart: DtStart? = null
    var dtEnd: DtEnd? = null

    var duration: Duration? = null
    val rRules = LinkedList<RRule>()
    val exRules = LinkedList<ExRule>()
    val rDates = LinkedList<RDate>()
    val exDates = LinkedList<ExDate>()

    val exceptions = LinkedList<Event>()

    var classification: Clazz? = null
    var status: Status? = null

    var opaque = true

    var organizer: Organizer? = null
    val attendees = LinkedList<Attendee>()

    val alarms = LinkedList<VAlarm>()

    var lastModified: LastModified? = null

    val categories = LinkedList<String>()
    val unknownProperties = LinkedList<Property>()

    companion object {
        /**
         * Parses an iCalendar resource, applies [ICalPreprocessor] to increase compatibility
         * and extracts the VEVENTs.
         *
         * @param reader where the iCalendar is taken from
         * @param properties Known iCalendar properties (like [CALENDAR_NAME]) will be put into this map. Key: property name; value: property value
         *
         * @return array of filled [Event] data objects (may have size 0)
         *
         * @throws ParserException when the iCalendar can't be parsed
         * @throws IllegalArgumentException when the iCalendar resource contains an invalid value
         * @throws IOException on I/O errors
         * @throws InvalidCalendarException on parsing exceptions
         */
        @UsesThreadContextClassLoader
        fun eventsFromReader(reader: Reader, properties: MutableMap<String, String>? = null): List<Event> {
            val ical = fromReader(reader, properties)

            // process VEVENTs
            val vEvents = ical.getComponents<VEvent>(Component.VEVENT)

            // make sure every event has an UID
            for (vEvent in vEvents)
                if (vEvent.uid == null) {
                    val uid = Uid(UUID.randomUUID().toString())
                    Ical4Android.log.warning("Found VEVENT without UID, using a random one: ${uid.value}")
                    vEvent.properties += uid
                }

            Ical4Android.log.fine("Assigning exceptions to main events")
            val mainEvents = mutableMapOf<String /* UID */,VEvent>()
            val exceptions = mutableMapOf<String /* UID */,MutableMap<String /* RECURRENCE-ID */,VEvent>>()
            for (vEvent in vEvents) {
                val uid = vEvent.uid.value
                val sequence = vEvent.sequence?.sequenceNo ?: 0

                if (vEvent.recurrenceId == null) {
                    // main event (no RECURRENCE-ID)

                    // If there are multiple entries, compare SEQUENCE and use the one with higher SEQUENCE.
                    // If the SEQUENCE is identical, use latest version.
                    val event = mainEvents[uid]
                    if (event == null || (event.sequence != null && sequence >= event.sequence.sequenceNo))
                        mainEvents[uid] = vEvent

                } else {
                    // exception (RECURRENCE-ID)
                    var ex = exceptions[uid]
                    // first index level: UID
                    if (ex == null) {
                        ex = mutableMapOf()
                        exceptions[uid] = ex
                    }
                    // second index level: RECURRENCE-ID
                    val recurrenceID = vEvent.recurrenceId.value
                    val event = ex[recurrenceID]
                    if (event == null || (event.sequence != null && sequence >= event.sequence.sequenceNo))
                        ex[recurrenceID] = vEvent
                }
            }

            /* There may be UIDs which have only RECURRENCE-ID entries and not a main entry (for instance, a recurring
            event with an exception where the current user has been invited only to this exception. In this case,
            the UID will not appear in mainEvents but only in exceptions. */

            val events = mutableListOf<Event>()
            for ((uid, vEvent) in mainEvents) {
                val event = fromVEvent(vEvent)

                // assign exceptions to main event and then remove them from exceptions array
                exceptions.remove(uid)?.let { eventExceptions ->
                    event.exceptions.addAll(eventExceptions.values.map { fromVEvent(it) })
                }

                // make sure that exceptions have at least a SUMMARY
                event.exceptions.forEach { it.summary = it.summary ?: event.summary }

                events += event
            }

            for ((uid, onlyExceptions) in exceptions) {
                Ical4Android.log.info("UID $uid doesn't have a main event but only exceptions: $onlyExceptions")

                // create a fake main event from the first exception
                val fakeEvent = fromVEvent(onlyExceptions.values.first())
                fakeEvent.exceptions.addAll(onlyExceptions.values.map { fromVEvent(it) })

                events += fakeEvent
            }

            return events
        }

        private fun fromVEvent(event: VEvent): Event {
            val e = Event()

            // sequence must only be null for locally created, not-yet-synchronized events
            e.sequence = 0

            // process properties
            for (prop in event.properties)
                when (prop) {
                    is Uid -> e.uid = prop.value
                    is RecurrenceId -> e.recurrenceId = prop
                    is Sequence -> e.sequence = prop.sequenceNo
                    is Summary -> e.summary = prop.value
                    is Location -> e.location = prop.value
                    is Url -> e.url = prop.uri
                    is Description -> e.description = prop.value
                    is Categories ->
                        for (category in prop.categories)
                            e.categories += category
                    is Color -> e.color = Css3Color.fromString(prop.value)
                    is DtStart -> e.dtStart = prop
                    is DtEnd -> e.dtEnd = prop
                    is Duration -> e.duration = prop
                    is RRule -> e.rRules += prop
                    is RDate -> e.rDates += prop
                    is ExRule -> e.exRules += prop
                    is ExDate -> e.exDates += prop
                    is Clazz -> e.classification = prop
                    is Status -> e.status = prop
                    is Transp -> e.opaque = prop == Transp.OPAQUE
                    is Organizer -> e.organizer = prop
                    is Attendee -> e.attendees += prop
                    is LastModified -> e.lastModified = prop
                    is ProdId, is DtStamp -> { /* don't save these as unknown properties */ }
                    else -> e.unknownProperties += prop
                }

            e.alarms.addAll(event.alarms)

            // validate and repair
            EventValidator(e).repair()

            return e
        }
    }


    @UsesThreadContextClassLoader
    fun write(os: OutputStream) {
        Ical4Android.checkThreadContextClassLoader()

        val ical = Calendar()
        ical.properties += Version.VERSION_2_0
        ical.properties +=
                if (userAgents.isEmpty())
                    prodId
                else
                    ProdId(prodId.value + " (" + userAgents.joinToString(",") + ")")

        val dtStart = dtStart ?: throw InvalidCalendarException("Won't generate event without start time")

        EventValidator(this).repair() // validate and repair this event before creating VEVENT

        // "main event" (without exceptions)
        val components = ical.components
        val mainEvent = toVEvent()
        components += mainEvent

        // remember used time zones
        val usedTimeZones = mutableSetOf<TimeZone>()
        dtStart.timeZone?.let(usedTimeZones::add)
        dtEnd?.timeZone?.let(usedTimeZones::add)

        // recurrence exceptions
        for (exception in exceptions) {
            // exceptions must always have the same UID as the main event
            exception.uid = uid

            val recurrenceId = exception.recurrenceId
            if (recurrenceId == null) {
                Ical4Android.log.warning("Ignoring exception without recurrenceId")
                continue
            }

            /* Exceptions must always have the same value type as DTSTART [RFC 5545 3.8.4.4].
               If this is not the case, we don't add the exception to the event because we're
               strict in what we send (and servers may reject such a case).
             */
            if (isDateTime(recurrenceId) != isDateTime(dtStart)) {
                Ical4Android.log.warning("Ignoring exception $recurrenceId with other date type than dtStart: $dtStart")
                continue
            }

            // for simplicity and compatibility, rewrite date-time exceptions to the same time zone as DTSTART
            if (isDateTime(recurrenceId) && recurrenceId.timeZone != dtStart.timeZone) {
                Ical4Android.log.fine("Changing timezone of $recurrenceId to same time zone as dtStart: $dtStart")
                recurrenceId.timeZone = dtStart.timeZone
            }

            // create and add VEVENT for exception
            val vException = exception.toVEvent()
            components += vException

            // remember used time zones
            exception.dtStart?.timeZone?.let(usedTimeZones::add)
            exception.dtEnd?.timeZone?.let(usedTimeZones::add)
        }

        // determine first dtStart (there may be exceptions with an earlier DTSTART that the main event)
        val dtStarts = mutableListOf(dtStart.date)
        dtStarts.addAll(exceptions.mapNotNull { it.dtStart?.date })
        val earliest = dtStarts.minOrNull()
        // add VTIMEZONE components
        for (tz in usedTimeZones)
            ical.components += minifyVTimeZone(tz.vTimeZone, earliest)

        softValidate(ical)
        CalendarOutputter(false).output(ical, os)
    }

    /**
     * Generates a VEvent representation of this event.
     *
     * @return generated VEvent
     */
    private fun toVEvent(): VEvent {
        val event = VEvent(/* generates DTSTAMP */)
        val props = event.properties
        props += Uid(uid)

        recurrenceId?.let { props += it }
        sequence?.let {
            if (it != 0)
                props += Sequence(it)
        }

        summary?.let { props += Summary(it) }
        location?.let { props += Location(it) }
        url?.let { props += Url(it) }
        description?.let { props += Description(it) }
        color?.let { props += Color(null, it.name) }

        dtStart?.let { props += it }
        dtEnd?.let { props += it }
        duration?.let { props += it }

        props.addAll(rRules)
        props.addAll(rDates)
        props.addAll(exRules)
        props.addAll(exDates)

        classification?.let { props += it }
        status?.let { props += it }
        if (!opaque)
            props += Transp.TRANSPARENT

        organizer?.let { props += it }
        props.addAll(attendees)

        if (categories.isNotEmpty())
            props += Categories(TextList(categories.toTypedArray()))
        props.addAll(unknownProperties)

        lastModified?.let { props += it }

        event.alarms.addAll(alarms)

        return event
    }


    val organizerEmail: String?
    get() {
        var email: String? = null
        organizer?.let { organizer ->
            val uri = organizer.calAddress
            email = if (uri.scheme.equals("mailto", true))
                uri.schemeSpecificPart
            else
                organizer.getParameter<Email>(Parameter.EMAIL)?.value
        }
        return email
    }

}
